codejail/proxy.py

```diff
@@ -0,0 +1,92 @@
+"""A proxy subprocess-making process for CodeJail."""
+
+import ast
+import os.path
+import subprocess
+import sys
+
+from .subproc import run_subprocess
+
+# We use .readline to get data from the pipes between the processes, so we need
+# to ensure that a newline does not appear in the data.  We also need a way to
+# communicate a few values, and unpack them.  Lastly, we need to be sure we can
+# handle binary data.  Serializing with repr() and deserializing the literals
+# that result give us all the properties we need.
+serialize = repr
+deserialize = ast.literal_eval
+
+# There is one global proxy process.
+PROXY_PROCESS = None
+
+def run_subprocess_through_proxy(*args, **kwargs):
+    """
+    Works just like :ref:`run_subprocess`, but through the proxy process.
+    """
+    global PROXY_PROCESS
+    if PROXY_PROCESS is None:
+        # Start the proxy by invoking proxy_main.py in our root directory.
+        root = os.path.dirname(os.path.dirname(os.path.realpath(__file__)))
+        proxy_main_py = os.path.join(root, "proxy_main.py")
+
+        # Run proxy_main.py with the same Python that is running us. "-u" makes
+        # the stdin and stdout unbuffered.
+        PROXY_PROCESS = subprocess.Popen(
+            [sys.executable, '-u', proxy_main_py],
+            bufsize=1,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            )
+
+    stdin = serialize((args, kwargs))
+    PROXY_PROCESS.stdin.write(stdin+"\n")
+
+    stdout = PROXY_PROCESS.stdout.readline()
+    retval = deserialize(stdout.rstrip())
+    return retval
+
+# Set this to True to see what the proxy process is doing.
+LOG_PROXY = False
+
+
+def proxy_main():
+    """
+    The main program for the proxy process.
+
+    It does this:
+
+        * Reads a JSON object from stdin: [args, kwargs]
```

Comment:
Eh, it doesn't really read a JSON object, it reads an object's `repr`. I think the JSON reference here is misleading. While Python reprs and JSON are similar, they have some syntactic differences and not all reprs are valid JSON.